Undo/COUndoTrackStore.m

```diff
@@ -283,55 +402,180 @@ - (id) deserialize: (NSData *)data
 {
-	[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	assert(dispatch_get_current_queue() != _queue);
+	
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	});
 }
 
 - (BOOL) string: (NSString *)aString matchesGlobPattern: (NSString *)aPattern
 {
-	return [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	assert(dispatch_get_current_queue() != _queue);
+	__block BOOL result = NO;
+
+	dispatch_sync(_queue, ^() {
+		result = [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	});
+	
+	return result;
+}
+
+- (void)markCommandsAsDeletedForUUIDs: (NSArray *)UUIDs
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	
+	__block NSArray *compactedTrackNames = nil;
+
+	// This can be run in background
+	dispatch_sync(_queue, ^() {
+		[_db beginTransaction];
+
+		for (ETUUID *UUID in UUIDs)
+		{
+			[_db executeUpdate: @"UPDATE commands SET deleted = 1 WHERE uuid = ?", [UUID dataValue]];
+		}
+
+		compactedTrackNames =
+			[_db arrayForQuery: @"SELECT DISTINCT trackname FROM commands WHERE deleted = 1 "];
+
+		[_db commit];
+	});
+
+	/* This must be run in the main thread:
+	   - no one must access _modifiedTrackStateForTrackName at the same time
+	   - we must post the commit notification immediately
+	   
+	   If we post it with some delay, someone could touch 
+	   _modifiedTrackStateForTrackName and overwrite COUndoTrackState.compacted. */
+	dispatch_sync_now(dispatch_get_main_queue(), ^() {
+		dispatch_sync(_queue, ^() {
+			[_db beginTransaction];
+
+			for (NSString *trackName in compactedTrackNames)
+			{
+				if (_modifiedTrackStateForTrackName[trackName] == nil)
+				{
+					_modifiedTrackStateForTrackName[trackName] = [self stateForTrackNameInCurrentQueue: trackName];
+				}
+				((COUndoTrackState *)_modifiedTrackStateForTrackName[trackName]).compacted = YES;
+			}
+
+			[_db commit];
+		});
+
+		[self postCommitNotifications];
+	});
+
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+- (void)finalizeDeletions
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE deleted = 1"];
+	});
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+/**
+ * We run vacuum with dispatch_sync() in a queue, so we can sure no other
+ * threads access the database when we invoke -vacuum in a background thread.
+ */
+- (BOOL)vacuum
+{
+    assert(dispatch_get_current_queue() != _queue);
+	__block BOOL success = NO;
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
```

Comment:
_transactionLock prevents the following sequence to be created in the database queue:

- begin transaction
- vacuum (or -markCommandsAsDeleted)
- end transaction (aka commit)

Since a COUndoTrackStore transaction is not submitted as a single block to the database queue, it seems to me we must prevent incompatible operations (e.g. vacuum or mark as deleted) to be submitted before this transaction ends.

_queue is used to run -vacuum in background (and blocks any other COUndoTrackStore operations that get submitted to the queue) while keeping the main thread available present a spinning indicator.
Beside this, there is no real concurrency use with _queue currently.

dispatch_async() doesn't seem compatible with -vacuum, since no other SQLite operations must be run during a vacuum, so I think we want to block the database queue during a vacuum operation. Am I missing something?